File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/crafting.ts
/**
 * Downtime crafting projects: characters spend days advancing a crafting
 * project toward completion. When a project completes, the crafted item is
 * added to the campaign's party inventory.
 */

import { getDb } from '../db.ts';
import type { ApiResult, JsonValue } from '../types.ts';
import { isValidIntInRange } from '../validation.ts';

interface CraftingProjectRow {
  id: string;
  character_id: string;
  item_slug: string;
  days_required: number;
  days_completed: number;
  status: string;
}

export function createCraftingProject(campaignId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const id = body.id;
  if (typeof id !== 'string' || id.length === 0) {
    return { status: 400, body: { error: 'id must be a non-empty string' } };
  }

  const characterId = body.character_id;
  if (typeof characterId !== 'string' || characterId.length === 0) {
    return { status: 400, body: { error: 'character_id must be a non-empty string' } };
  }

  const character = db
    .prepare('SELECT id FROM campaign_characters WHERE campaign_id = ? AND id = ?')
    .get(campaignId, characterId);
  if (!character) {
    return { status: 404, body: { error: 'unknown character id' } };
  }

  const itemSlug = body.item_slug;
  if (typeof itemSlug !== 'string' || itemSlug.length === 0) {
    return { status: 400, body: { error: 'item_slug must be a non-empty string' } };
  }

  const daysRequired = body.days_required;
  if (!isValidIntInRange(daysRequired, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'days_required must be a positive integer' } };
  }

  const costGp = body.cost_gp;
  if (typeof costGp !== 'number' || !Number.isFinite(costGp) || costGp < 0) {
    return { status: 400, body: { error: 'cost_gp must be a non-negative number' } };
  }

  const existing = db
    .prepare('SELECT id FROM campaign_crafting_projects WHERE campaign_id = ? AND id = ?')
    .get(campaignId, id);
  if (existing) {
    return { status: 409, body: { error: 'crafting project id already exists' } };
  }

  db.prepare(
    `INSERT INTO campaign_crafting_projects
       (campaign_id, id, character_id, item_slug, days_required, days_completed, cost_gp, status)
     VALUES (?, ?, ?, ?, ?, 0, ?, 'active')`,
  ).run(campaignId, id, characterId, itemSlug, daysRequired, costGp);

  return {
    status: 201,
    body: {
      id,
      character_id: characterId,
      item_slug: itemSlug,
      days_required: daysRequired,
      days_completed: 0,
      status: 'active',
    },
  };
}

export function advanceCrafting(campaignId: string, projectId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const project = db
    .prepare(
      'SELECT id, character_id, item_slug, days_required, days_completed, status FROM campaign_crafting_projects WHERE campaign_id = ? AND id = ?',
    )
    .get(campaignId, projectId) as CraftingProjectRow | undefined;
  if (!project) {
    return { status: 404, body: { error: 'unknown crafting project id' } };
  }

  const days = body.days;
  if (!isValidIntInRange(days, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'days must be a positive integer' } };
  }

  if (project.status === 'complete') {
    return { status: 409, body: { error: 'crafting project is already complete' } };
  }

  const daysCompleted = Math.min(project.days_required, project.days_completed + days);
  const status = daysCompleted >= project.days_required ? 'complete' : 'active';

  db.prepare(
    'UPDATE campaign_crafting_projects SET days_completed = ?, status = ? WHERE campaign_id = ? AND id = ?',
  ).run(daysCompleted, status, campaignId, projectId);

  if (status === 'complete' && project.status !== 'complete') {
    db.prepare(
      `INSERT INTO campaign_inventory (campaign_id, item_slug, owner, quantity)
       VALUES (?, ?, 'party', 1)
       ON CONFLICT (campaign_id, item_slug, owner) DO UPDATE SET quantity = quantity + excluded.quantity`,
    ).run(campaignId, project.item_slug);
  }

  return {
    status: 200,
    body: {
      id: project.id,
      days_completed: daysCompleted,
      status,
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075409Z_pi_kimi-k2p7-code_ruby-stdlib/lib/crafting.rb
# frozen_string_literal: true

require_relative 'persistence'

# Downtime crafting: long-term item crafting projects.
module Crafting
  VALID_STATUSES = %w[active complete].freeze

  def self.create_project(campaign_id, payload)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty?

    data = validate_project(payload)
    return [:invalid] unless data

    Persistence.db do |d|
      campaign_exists = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless campaign_exists

      character_exists = d.get_first_value(
        'SELECT 1 FROM campaign_characters WHERE id = ? AND campaign_id = ?',
        [data[:character_id], campaign_id]
      )
      next [:not_found] unless character_exists

      existing = d.get_first_value('SELECT 1 FROM campaign_crafting_projects WHERE id = ?', data[:id])
      next [:conflict] if existing

      d.execute(
        'INSERT INTO campaign_crafting_projects (id, campaign_id, character_id, item_slug, days_required, days_completed, cost_gp, status) VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
        [data[:id], campaign_id, data[:character_id], data[:item_slug], data[:days_required], 0, data[:cost_gp], 'active']
      )

      [:ok, {
        'id' => data[:id],
        'character_id' => data[:character_id],
        'item_slug' => data[:item_slug],
        'days_required' => data[:days_required],
        'days_completed' => 0,
        'status' => 'active'
      }]
    end
  end

  def self.advance_project(campaign_id, project_id, payload)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty? &&
                             project_id.is_a?(String) && !project_id.empty?

    days = payload['days']
    return [:invalid] unless days.is_a?(Integer) && days > 0

    Persistence.db do |d|
      row = d.get_first_row(
        'SELECT id, days_required, days_completed, status, item_slug FROM campaign_crafting_projects WHERE id = ? AND campaign_id = ?',
        [project_id, campaign_id]
      )
      next [:not_found] unless row

      project_id, days_required, days_completed, status, item_slug = row

      if status == 'complete'
        next [:ok, {
          'id' => project_id,
          'days_completed' => days_completed,
          'status' => 'complete'
        }]
      end

      new_days_completed = [days_completed + days, days_required].min
      new_status = new_days_completed >= days_required ? 'complete' : 'active'

      if new_status == 'complete'
        d.execute(
          'INSERT INTO campaign_inventory (campaign_id, item_slug, owner, quantity) VALUES (?, ?, ?, ?)
           ON CONFLICT(campaign_id, item_slug, owner) DO UPDATE SET quantity = quantity + excluded.quantity',
          [campaign_id, item_slug, 'party', 1]
        )
      end

      d.execute(
        'UPDATE campaign_crafting_projects SET days_completed = ?, status = ? WHERE id = ?',
        [new_days_completed, new_status, project_id]
      )

      [:ok, {
        'id' => project_id,
        'days_completed' => new_days_completed,
        'status' => new_status
      }]
    end
  end

  def self.validate_project(payload)
    return nil unless payload.is_a?(Hash)

    id = payload['id']
    character_id = payload['character_id']
    item_slug = payload['item_slug']
    days_required = payload['days_required']
    cost_gp = payload['cost_gp']

    return nil unless id.is_a?(String) && !id.empty? &&
                      character_id.is_a?(String) && !character_id.empty? &&
                      item_slug.is_a?(String) && !item_slug.empty? &&
                      days_required.is_a?(Integer) && days_required > 0 &&
                      cost_gp.is_a?(Integer)

    {
      id: id,
      character_id: character_id,
      item_slug: item_slug,
      days_required: days_required,
      cost_gp: cost_gp
    }
  end
  private_class_method :validate_project
end
